README.md
# simpleStorage

Cross-browser key-value store database to store data locally in the browser.

*simpleStorage* is a fork of [jStorage](http://www.jstorage.info/) that only includes the minimal set of features. Basically it is a wrapper for native `JSON` + `localStorage` with some TTL magic mixed in.

The module has no dependencies, you can use it as a standalone script (introduces `simpleStorage` global) or as an AMD module. All modern browsers (including mobile) are supported, older browsers (IE7, Firefox 3) are not.

*simpleStorage* is **very** small - about **1kB** in size when minimized and gzipped.

[![Build Status](https://travis-ci.org/andris9/simpleStorage.png?branch=master)](https://travis-ci.org/andris9/simpleStorage)

## Install

Quickest way to get up and running woulr be to use [jsDelivr CDN](http://www.jsdelivr.com/projects/simplestorage):

```html
<script src="https://cdn.jsdelivr.net/simplestorage/0.2.1/simpleStorage.min.js"></script>
```

Otherwise you can download [simpleStorage.js](https://github.com/andris9/simpleStorage/blob/master/simpleStorage.js) or install it with bower:

    bower install simpleStorage

and include the following script in your web application: *bower_components/simpleStorage/simpleStorage.js*

or install with npm

    npm install simplestorage.js

## Support simpleStorage development

[![Donate to author](https://www.paypalobjects.com/en_US/i/btn/btn_donate_SM.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=DB26KWR2BQX5W)

If you want to support with Bitcoins, then my wallet address is `15Z8ADxhssKUiwP3jbbqJwA21744KMCfTM`

## Usage

*simpleStorage* API is a subset of jStorage with slight modifications, so for most cases it should work out of the box if you are converting from jStorage. Main difference is between return values - if an action failed because of an error (storage full, storage not available, invalid data used etc.), you get the error object as the return value. jStorage never indicated anything if an error occurred.

Possible error codes (from `err.code`):

  * **"LS_NOT_AVAILABLE"** means that localStorage is not supported by this browser
  * **"LS_DISABLE"** means that localStorage is supported by this browser but it can't be used for whatever reason (privacy mode, manual disabling etc.)
  * **"LS_QUOTA_EXCEEDED"** means that the allocated quota for localStorage is all used up or would be if current value is stored
  * *anything else*, no idea

### set(key, value[, options])

Store or update a value in local storage.

```javascript
simpleStorage.set(key, value[, options])
```

Where

  * **key** - the key for the value
  * **value** - value to be stored (can be any JSONeable value)
  * **options** - optional options object. Currently the only available option is `TTL` which sets the time-to-live (TTL) value in milliseconds for the given key/value

```javascript
// the following entry expires in 100 seconds
simpleStorage.set(key, value, {TTL: 100000})
```

Return values

  * **true** - value was stored
  * **false** - value was not stored
  * Error object - value was not stored because of an error. See `error.code` for explanation

### get(key)

Retrieve a value from local storage.

```javascript
value = simpleStorage.get(key)
```

Where

  * **key** - the key to be retrieved

Method returns the value for a key or undefined if the key was not found.

### hasKey(key)

Checks if there's a value with the given key in the local storage.

```javascript
simpleStorage.hasKey(key)
```

Where

  * **key** - the key to be checked

Method returns true if the given key exists, false otherwise.

### deleteKey(key)

Removes a value from local storage.

```javascript
simpleStorage.deleteKey(key)
```

Return values

  * **true** - value was deleted
  * **false** - value was not found
  * Error object - value was not deleted because of an error. See `error.code` for explanation

### setTTL(key, ttl)

Set a millisecond timeout. When the timeout is reached, the key is removed automatically from local storage.

```javascript
simpleStorage.setTTL(key, ttl)
```

Where

  * **key** - the key to be updated
  * **ttl** - timeout in milliseconds. If the value is 0, timeout is cleared from the key

Return values

  * **true** - ttl was set
  * **false** - value was not found
  * Error object - ttl was not set because of an error. See `error.code` for explanation

### getTTL(key)

Retrieve remaining milliseconds for a key with TTL

```javascript
ttl = simpleStorage.getTTL(key)
```

Where

  * **key** - the key to be checked

Return values

  * finite **number** - remaining milliseconds
  * **Infinity** - TTL is not set for the selected key
  * **false** - selected key does not exist or is expired

### flush()

Clear all values

```javascript
simpleStorage.flush()
```

Return values

  * **true** - storage was flushed
  * Error object - storage was not flushed because of an error. See `error.code` for explanation

### index()

Retrieve all used keys as an array

```javascript
list = simpleStorage.index()
```

Returns an array of keys.

### storageSize()

Get used storage in symbol count

```javascript
simpleStorage.storageSize()
```

### canUse()

Check if local storage can be used

```javascript
simpleStorage.canUse()
```

Returns true if storage is available

## Demo

See demo [here](http://tahvel.info/simpleStorage/example/).

## Tests

See QUnit tests [here](http://tahvel.info/simpleStorage/tests/).

## License

[Unlicense](http://unlicense.org/)


CHANGELOG.md
# Changelog

## v0.2.1 2016-02-06

  * Added new method `hasKey(key)`

## v0.2.0 2015-11-06

  * Unified error handling for over quota, disabled local storage etc.

## v0.1.4 2015-10-27

  * Added Gruntfile to auto generate simpleStorage.min.js
  * Added jshint rule file and Grunt task

## v0.1.3 2015-01-15

  * Bumped version to 0.1.3
  * Changed formatting [68d600b4]
  * #18 fix duplicate keys (will-e-yams) [28a1bd2b]

## v0.1.2 2014-02-16

  * Bumped version to 0.1.2
  * Refactored TTL cleanup [1e176925]
  * Added travis-ci config for auto running tests after every commit [845df28a]

## v0.1.1 2014-02-15

  * Bumped version to 0.1.1
  * Added bower package file [a3faa6b5]
  * getTTL returns a number (TTL is set), Infinity (TTL is not set) or false (key does not exist) [36800824]

## v0.1.0 2014-01-11

  * Initial release


tests/tests.js
/* jshint browser: true */
/* global QUnit: false, simpleStorage: false, asyncTest: false, expect: false,
   start: false, test: false, ok: false, deepEqual: false, equal: false */

QUnit.testStart(function() {
    "use strict";

    simpleStorage.flush();
});

QUnit.testDone(function() {
    "use strict";

    simpleStorage.flush();
});

module("general");

test("can use", function() {
    "use strict";

    ok(simpleStorage.canUse());
});

test("can not use", function() {
    "use strict";

    ok(simpleStorage.canUse());
});

test("flush/index", function() {
    "use strict";

    equal(simpleStorage.set("test", "value"), true);
    deepEqual(simpleStorage.index(), ["test"]);
    equal(simpleStorage.flush(), true);
    deepEqual(simpleStorage.index(), []);
    ok(!simpleStorage.get("test"));
});

module("keys");

test("missing", function() {
    "use strict";

    ok(typeof simpleStorage.get("test") == "undefined");
});

test("string", function() {
    "use strict";

    equal(simpleStorage.set("test", "value"), true);
    ok(simpleStorage.get("test") == "value");
});

test("boolean", function() {
    "use strict";

    equal(simpleStorage.set("test true", true), true);
    ok(simpleStorage.get("test true") === true);
    equal(simpleStorage.set("test false", false), true);
    ok(simpleStorage.get("test false") === false);
});

test("number", function() {
    "use strict";

    equal(simpleStorage.set("test", 10.01), true);
    ok(simpleStorage.get("test") === 10.01);
});

test("obejct", function() {
    "use strict";

    var testObj = {arr:[1,2,3]};
    equal(simpleStorage.set("test", testObj), true);
    deepEqual(simpleStorage.get("test"), testObj);
    ok(simpleStorage.get("test") != testObj);
});

test("deleteKey", function() {
    "use strict";

    deepEqual(simpleStorage.index(), []);
    equal(simpleStorage.set("test", "value"), true);
    deepEqual(simpleStorage.index(), ["test"]);
    equal(simpleStorage.deleteKey("test"), true);
    equal(simpleStorage.deleteKey("test"), false);
    deepEqual(simpleStorage.index(), []);
});

module("ttl");

asyncTest("TTL", function() {
    "use strict";

    expect(3);
    equal(simpleStorage.set("ttlkey", "value", {TTL: 500}), true);
    setTimeout(function(){
      ok(simpleStorage.get("ttlkey") == "value");
      setTimeout(function(){
          ok(typeof simpleStorage.get("ttlkey") == "undefined");
          start();
      }, 500);
    }, 250);
});

asyncTest("setTTL", function() {
    "use strict";

    expect(4);
    equal(simpleStorage.set("ttlkey", "value"), true);
    equal(simpleStorage.setTTL("ttlkey", 500), true);
    setTimeout(function(){
        ok(simpleStorage.get("ttlkey") == "value");
        setTimeout(function(){
            ok(typeof simpleStorage.get("ttlkey") == "undefined");
            start();
        }, 500);
    }, 250);
});

asyncTest("setTTL, negative value", function() {
    "use strict";

    expect(3);
    equal(simpleStorage.set("ttlkey", "value"), true);
    equal(simpleStorage.setTTL("ttlkey", -1), true);
    setTimeout(function(){
        ok(typeof simpleStorage.get("ttlkey") == "undefined");
        start();
    }, 250);
});

asyncTest("getTTL", function() {
    "use strict";

    expect(6);
    equal(simpleStorage.set("ttlkey", "value"), true);
    equal(simpleStorage.getTTL("ttlkey"), Infinity);
    equal(simpleStorage.setTTL("ttlkey", 500), true);
    setTimeout(function(){
        ok(simpleStorage.getTTL("ttlkey") > 0);
        setTimeout(function(){
            equal(simpleStorage.getTTL("ttlkey"), false);
            ok(typeof simpleStorage.get("ttlkey") == "undefined");
            start();
      }, 500);
    }, 250);
});

asyncTest("Clearing TTL removes meta object", function() {
    "use strict";

    expect(4);
    ok(!localStorage.simpleStorage);
    equal(simpleStorage.set("ttlkey", "value", {TTL: 100}), true);
    ok(JSON.parse(localStorage.simpleStorage).__simpleStorage_meta);
    setTimeout(function(){
        ok(!JSON.parse(localStorage.simpleStorage).__simpleStorage_meta);
        start();
    }, 200);
});


Gruntfile.js
/* jshint node: true */

'use strict';

module.exports = function(grunt) {

    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),

        jshint: {
            all: ['test/**/*.js', 'Gruntfile.js', 'simpleStorage.js'],
            options: {
                jshintrc: '.jshintrc'
            }
        },

        uglify: {
            options: {
                banner: '/*! simpleStorage v<%= pkg.version %>, Unlicense <%= grunt.template.today("yyyy") %>. https://github.com/andris9/simpleStorage */\n'
            },
            build: {
                src: 'simpleStorage.js',
                dest: 'simpleStorage.min.js'
            }
        }
    });

    grunt.loadNpmTasks('grunt-contrib-jshint');
    grunt.loadNpmTasks('grunt-contrib-uglify');

    grunt.registerTask('default', ['jshint', 'uglify']);

};


simpleStorage.js
/* jshint browser: true */
/* global define: false, module: false */

// AMD shim
(function(root, factory) {

    'use strict';

    if (typeof define === 'function' && define.amd) {
        define(factory);
    } else if (typeof exports !== 'undefined') {
        module.exports = factory();
    } else {
        root.simpleStorage = factory();
    }

}(this, function() {

    'use strict';

    var VERSION = '0.2.0';

    /* This is the object, that holds the cached values */
    var _storage = false;

    /* How much space does the storage take */
    var _storage_size = 0;

    var _storage_available = false;
    var _ttl_timeout = null;

    var _lsStatus = 'OK';
    var LS_NOT_AVAILABLE = 'LS_NOT_AVAILABLE';
    var LS_DISABLED = 'LS_DISABLED';
    var LS_QUOTA_EXCEEDED = 'LS_QUOTA_EXCEEDED';

    // This method might throw as it touches localStorage and doing so
    // can be prohibited in some environments
    function _init() {

        // this method throws if localStorage is not usable, otherwise returns true
        _storage_available = _checkAvailability();

        // Load data from storage
        _loadStorage();

        // remove dead keys
        _handleTTL();

        // start listening for changes
        _setupUpdateObserver();

        // handle cached navigation
        if ('addEventListener' in window) {
            window.addEventListener('pageshow', function(event) {
                if (event.persisted) {
                    _reloadData();
                }
            }, false);
        }

        _storage_available = true;
    }

    /**
     * Sets up a storage change observer
     */
    function _setupUpdateObserver() {
        if ('addEventListener' in window) {
            window.addEventListener('storage', _reloadData, false);
        } else {
            document.attachEvent('onstorage', _reloadData);
        }
    }

    /**
     * Reload data from storage when needed
     */
    function _reloadData() {
        try {
            _loadStorage();
        } catch (E) {
            _storage_available = false;
            return;
        }
        _handleTTL();
    }

    function _loadStorage() {
        var source = localStorage.getItem('simpleStorage');

        try {
            _storage = JSON.parse(source) || {};
        } catch (E) {
            _storage = {};
        }

        _storage_size = _get_storage_size();
    }

    function _save() {
        try {
            localStorage.setItem('simpleStorage', JSON.stringify(_storage));
            _storage_size = _get_storage_size();
        } catch (E) {
            return _formatError(E);
        }
        return true;
    }

    function _get_storage_size() {
        var source = localStorage.getItem('simpleStorage');
        return source ? String(source).length : 0;
    }

    function _handleTTL() {
        var curtime, i, len, expire, keys, nextExpire = Infinity,
            expiredKeysCount = 0;

        clearTimeout(_ttl_timeout);

        if (!_storage || !_storage.__simpleStorage_meta || !_storage.__simpleStorage_meta.TTL) {
            return;
        }

        curtime = +new Date();
        keys = _storage.__simpleStorage_meta.TTL.keys || [];
        expire = _storage.__simpleStorage_meta.TTL.expire || {};

        for (i = 0, len = keys.length; i < len; i++) {
            if (expire[keys[i]] <= curtime) {
                expiredKeysCount++;
                delete _storage[keys[i]];
                delete expire[keys[i]];
            } else {
                if (expire[keys[i]] < nextExpire) {
                    nextExpire = expire[keys[i]];
                }
                break;
            }
        }

        // set next check
        if (nextExpire !== Infinity) {
            _ttl_timeout = setTimeout(_handleTTL, Math.min(nextExpire - curtime, 0x7FFFFFFF));
        }

        // remove expired from TTL list and save changes
        if (expiredKeysCount) {
            keys.splice(0, expiredKeysCount);

            _cleanMetaObject();
            _save();
        }
    }

    function _setTTL(key, ttl) {
        var curtime = +new Date(),
            i, len, added = false;

        ttl = Number(ttl) || 0;

        // Set TTL value for the key
        if (ttl !== 0) {
            // If key exists, set TTL
            if (_storage.hasOwnProperty(key)) {

                if (!_storage.__simpleStorage_meta) {
                    _storage.__simpleStorage_meta = {};
                }

                if (!_storage.__simpleStorage_meta.TTL) {
                    _storage.__simpleStorage_meta.TTL = {
                        expire: {},
                        keys: []
                    };
                }

                _storage.__simpleStorage_meta.TTL.expire[key] = curtime + ttl;

                // find the expiring key in the array and remove it and all before it (because of sort)
                if (_storage.__simpleStorage_meta.TTL.expire.hasOwnProperty(key)) {
                    for (i = 0, len = _storage.__simpleStorage_meta.TTL.keys.length; i < len; i++) {
                        if (_storage.__simpleStorage_meta.TTL.keys[i] === key) {
                            _storage.__simpleStorage_meta.TTL.keys.splice(i);
                        }
                    }
                }

                // add key to keys array preserving sort (soonest first)
                for (i = 0, len = _storage.__simpleStorage_meta.TTL.keys.length; i < len; i++) {
                    if (_storage.__simpleStorage_meta.TTL.expire[_storage.__simpleStorage_meta.TTL.keys[i]] > (curtime + ttl)) {
                        _storage.__simpleStorage_meta.TTL.keys.splice(i, 0, key);
                        added = true;
                        break;
                    }
                }

                // if not added in previous loop, add here
                if (!added) {
                    _storage.__simpleStorage_meta.TTL.keys.push(key);
                }
            } else {
                return false;
            }
        } else {
            // Remove TTL if set
            if (_storage && _storage.__simpleStorage_meta && _storage.__simpleStorage_meta.TTL) {

                if (_storage.__simpleStorage_meta.TTL.expire.hasOwnProperty(key)) {
                    delete _storage.__simpleStorage_meta.TTL.expire[key];
                    for (i = 0, len = _storage.__simpleStorage_meta.TTL.keys.length; i < len; i++) {
                        if (_storage.__simpleStorage_meta.TTL.keys[i] === key) {
                            _storage.__simpleStorage_meta.TTL.keys.splice(i, 1);
                            break;
                        }
                    }
                }

                _cleanMetaObject();
            }
        }

        // schedule next TTL check
        clearTimeout(_ttl_timeout);
        if (_storage && _storage.__simpleStorage_meta && _storage.__simpleStorage_meta.TTL && _storage.__simpleStorage_meta.TTL.keys.length) {
            _ttl_timeout = setTimeout(_handleTTL, Math.min(Math.max(_storage.__simpleStorage_meta.TTL.expire[_storage.__simpleStorage_meta.TTL.keys[0]] - curtime, 0), 0x7FFFFFFF));
        }

        return true;
    }

    function _cleanMetaObject() {
        var updated = false,
            hasProperties = false,
            i;

        if (!_storage || !_storage.__simpleStorage_meta) {
            return updated;
        }

        // If nothing to TTL, remove the object
        if (_storage.__simpleStorage_meta.TTL && !_storage.__simpleStorage_meta.TTL.keys.length) {
            delete _storage.__simpleStorage_meta.TTL;
            updated = true;
        }

        // If meta object is empty, remove it
        for (i in _storage.__simpleStorage_meta) {
            if (_storage.__simpleStorage_meta.hasOwnProperty(i)) {
                hasProperties = true;
                break;
            }
        }

        if (!hasProperties) {
            delete _storage.__simpleStorage_meta;
            updated = true;
        }

        return updated;
    }

    /**
     * Checks if localStorage is available or throws an error
     */
    function _checkAvailability() {
        var err;
        var items = 0;

        // Firefox sets localStorage to 'null' if support is disabled
        // IE might go crazy if quota is exceeded and start treating it as 'unknown'
        if (window.localStorage === null || typeof window.localStorage === 'unknown') {
            err = new Error('localStorage is disabled');
            err.code = LS_DISABLED;
            throw err;
        }

        // There doesn't seem to be any indication about localStorage support
        if (!window.localStorage) {
            err = new Error('localStorage not supported');
            err.code = LS_NOT_AVAILABLE;
            throw err;
        }

        try {
            items = window.localStorage.length;
        } catch (E) {
            throw _formatError(E);
        }

        try {
            // we try to set a value to see if localStorage is really usable or not
            window.localStorage.setItem('__simpleStorageInitTest', Date.now().toString(16));
            window.localStorage.removeItem('__simpleStorageInitTest');
        } catch (E) {
            if (items) {
                // there is already some data stored, so this might mean that storage is full
                throw _formatError(E);
            } else {
                // we do not have any data stored and we can't add anything new
                // so we are most probably in Private Browsing mode where
                // localStorage is turned off in some browsers (max storage size is 0)
                err = new Error('localStorage is disabled');
                err.code = LS_DISABLED;
                throw err;
            }
        }

        return true;
    }

    function _formatError(E) {
        var err;

        // No more storage:
        // Mozilla: NS_ERROR_DOM_QUOTA_REACHED, code 1014
        // WebKit: QuotaExceededError/QUOTA_EXCEEDED_ERR, code 22
        // IE number -2146828281: Out of memory
        // IE number -2147024882: Not enough storage is available to complete this operation
        if (E.code === 22 || E.code === 1014 || [-2147024882, -2146828281, -21474675259].indexOf(E.number) > 0) {
            err = new Error('localStorage quota exceeded');
            err.code = LS_QUOTA_EXCEEDED;
            return err;
        }

        // SecurityError, localStorage is turned off
        if (E.code === 18 || E.code === 1000) {
            err = new Error('localStorage is disabled');
            err.code = LS_DISABLED;
            return err;
        }

        // We are trying to access something from an object that is either null or undefined
        if (E.name === 'TypeError') {
            err = new Error('localStorage is disabled');
            err.code = LS_DISABLED;
            return err;
        }

        return E;
    }

    // Sets value for _lsStatus
    function _checkError(err) {
        if (!err) {
            _lsStatus = 'OK';
            return err;
        }

        switch (err.code) {
            case LS_NOT_AVAILABLE:
            case LS_DISABLED:
            case LS_QUOTA_EXCEEDED:
                _lsStatus = err.code;
                break;
            default:
                _lsStatus = err.code || err.number || err.message || err.name;
        }

        return err;
    }

    ////////////////////////// PUBLIC INTERFACE /////////////////////////

    try {
        _init();
    } catch (E) {
        _checkError(E);
    }

    return {

        version: VERSION,

        status: _lsStatus,

        canUse: function() {
            return _lsStatus === 'OK' && !!_storage_available;
        },

        set: function(key, value, options) {
            if (key === '__simpleStorage_meta') {
                return false;
            }

            if (!_storage) {
                return false;
            }

            // undefined values are deleted automatically
            if (typeof value === 'undefined') {
                return this.deleteKey(key);
            }

            options = options || {};

            // Check if the value is JSON compatible (and remove reference to existing objects/arrays)
            try {
                value = JSON.parse(JSON.stringify(value));
            } catch (E) {
                return _formatError(E);
            }

            _storage[key] = value;

            _setTTL(key, options.TTL || 0);

            return _save();
        },

        hasKey: function(key) {
            if (!_storage) {
                return false;
            }

            if (_storage.hasOwnProperty(key) && key !== '__simpleStorage_meta') {
                return true;
            }

            return false;
        },

        get: function(key) {
            if (!_storage) {
                return false;
            }

            if (_storage.hasOwnProperty(key) && key !== '__simpleStorage_meta') {
                // TTL value for an existing key is either a positive number or an Infinity
                if (this.getTTL(key)) {
                    return _storage[key];
                }
            }
        },

        deleteKey: function(key) {

            if (!_storage) {
                return false;
            }

            if (key in _storage) {
                delete _storage[key];

                _setTTL(key, 0);

                return _save();
            }

            return false;
        },

        setTTL: function(key, ttl) {
            if (!_storage) {
                return false;
            }

            _setTTL(key, ttl);

            return _save();
        },

        getTTL: function(key) {
            var ttl;

            if (!_storage) {
                return false;
            }

            if (_storage.hasOwnProperty(key)) {
                if (_storage.__simpleStorage_meta &&
                    _storage.__simpleStorage_meta.TTL &&
                    _storage.__simpleStorage_meta.TTL.expire &&
                    _storage.__simpleStorage_meta.TTL.expire.hasOwnProperty(key)) {

                    ttl = Math.max(_storage.__simpleStorage_meta.TTL.expire[key] - (+new Date()) || 0, 0);

                    return ttl || false;
                } else {
                    return Infinity;
                }
            }

            return false;
        },

        flush: function() {
            if (!_storage) {
                return false;
            }

            _storage = {};
            try {
                localStorage.removeItem('simpleStorage');
                return true;
            } catch (E) {
                return _formatError(E);
            }
        },

        index: function() {
            if (!_storage) {
                return false;
            }

            var index = [],
                i;
            for (i in _storage) {
                if (_storage.hasOwnProperty(i) && i !== '__simpleStorage_meta') {
                    index.push(i);
                }
            }
            return index;
        },

        storageSize: function() {
            return _storage_size;
        }
    };

}));
